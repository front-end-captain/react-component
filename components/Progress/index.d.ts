import * as React from "react";

export interface ProgressProps {
  // 所有样式类名前缀 默认 sky-progress
  prefixCls?: string;

  // 应用于包裹容器的样式类名
  className?: string;

  // 进度条类型 默认 line
  type?: "line" | "circle" | "dashboard";

  // 进度条占比
  percent?: number;

  // 是否显示提示内容 默认 true
  showInfo?: boolean;

  // 进度条状态 默认 normal
  status?: "normal" | "exception" | "success";

  // 进度条宽度 默认 1.5
  lineWidth?: number;

  // 进度条边缘的形状 默认 round
  linecap?: "round" | "square";

  // 进度条颜色 默认 #13c2c2
  lineColor?: string;

  // 应用于 svg 元素的样式
  style?: React.CSSProperties;

  // 应用于包裹容器的样式
  wrapStyle?: React.CSSProperties;

  // 指定提示内容的模板函数 默认 {percent}%
  format?: (percent?: number) => React.ReactNode;
}

declare const Progress: (ProgressProps) => React.ReactNode;

export default Progress;
