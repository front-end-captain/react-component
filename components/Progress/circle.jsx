import React from "react";
import classNames from "classnames";

import { CircleWrapper } from "./progress.css.js";

const Circle = (props) => {
  const getPathStylesAndPathString = (
    offset,
    percent,
    lineWidth,
    gapDegree = 0,
    gapPosition = "bottom",
  ) => {
    const radius = 50 - lineWidth / 2;

    let XCoordOfStartPoint = 0;
    let YCoordOfStartPoint = -radius;
    let XCoordOfEndPoint = 0;
    let YCoordOfEndPoint = -2 * radius;

    switch (gapPosition) {
      case "left":
        XCoordOfStartPoint = -radius;
        YCoordOfStartPoint = 0;
        XCoordOfEndPoint = 2 * radius;
        YCoordOfEndPoint = 0;
        break;
      case "right":
        XCoordOfStartPoint = radius;
        YCoordOfStartPoint = 0;
        XCoordOfEndPoint = -2 * radius;
        YCoordOfEndPoint = 0;
        break;
      case "bottom":
        YCoordOfStartPoint = radius;
        YCoordOfEndPoint = 2 * radius;
        break;
      default:
    }

    const pathString = `M 50,50 m ${XCoordOfStartPoint},${YCoordOfStartPoint}
     a ${radius},${radius} 0 1 1 ${XCoordOfEndPoint},${-YCoordOfEndPoint}
     a ${radius},${radius} 0 1 1 ${-XCoordOfEndPoint},${YCoordOfEndPoint}`;

    const perimeter = Math.PI * 2 * radius;

    const pathStyle = {
      strokeDasharray: `${(percent / 100) * (perimeter - gapDegree)}px ${perimeter}px`,
      strokeDashoffset: `-${gapDegree / 2 + (offset / 100) * (perimeter - gapDegree)}px`,
      transition:
        "stroke-dashoffset 0.3s ease 0s, stroke-dasharray 0.3s ease 0s, stroke 0.3s, stroke-width 0.06s ease 0.3s", // eslint-disable-line
    };

    return {
      pathString,
      pathStyle,
    };
  };

  const {
    prefixCls,
    className,
    style,
    wrapStyle,
    type,
    percent,
    linecap,
    lineWidth,
    format,
    showInfo,
    status,
    trailColor,
    adoptStrokeColor,
    adoptInfoContent,
  } = props;

  const realLineWidth = lineWidth * 3;
  const gapDegree = type === "dashboard" ? 70 : 0;
  const gapPosition = type === "dashboard" ? "bottom" : "top";

  const { pathString: trailPathString, pathStyle: trailPathStyle } = getPathStylesAndPathString(
    0,
    100,
    realLineWidth,
    gapDegree,
    gapPosition,
  );

  const { pathString: strokePathString, pathStyle: strokePathStyle } = getPathStylesAndPathString(
    0,
    percent,
    realLineWidth,
    gapDegree,
    gapPosition,
  );

  return (
    <CircleWrapper
      prefixCls={prefixCls}
      className={classNames(
        `${prefixCls}-circle-wrapper`,
        `${prefixCls}-circle-${status}`,
        className,
      )}
      style={wrapStyle}
    >
      <svg className={`${prefixCls}-circle`} viewBox="0 0 100 100" style={style}>
        <path
          className={`${prefixCls}-circle-trail`}
          d={trailPathString}
          stroke={trailColor}
          strokeLinecap={linecap}
          strokeWidth={realLineWidth}
          fillOpacity="0"
          style={trailPathStyle}
        />
        <path
          className={`${prefixCls}-circle-path`}
          d={strokePathString}
          stroke={adoptStrokeColor(percent, status)}
          strokeLinecap={linecap}
          strokeWidth={percent === 0 ? 0 : realLineWidth}
          fillOpacity="0"
          style={strokePathStyle}
        />
      </svg>
      {showInfo && (
        <span className={`${prefixCls}-info`}>{adoptInfoContent(percent, status, format)}</span>
      )}
    </CircleWrapper>
  );
};

export default Circle;
