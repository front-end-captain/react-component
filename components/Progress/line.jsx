import React from "react";
import classNames from "classnames";

import { LineWrapper } from "./progress.css.js";

const Line = (props) => {
  const {
    className,
    percent,
    prefixCls,
    linecap,
    lineWidth,
    format,
    style,
    showInfo,
    status,
    trailColor,
    wrapStyle,
    adoptStrokeColor,
    adoptInfoContent,
  } = props;

  const YCoordOfStartPoint = lineWidth / 2;
  const YCoordOfEndPoint = 100 - lineWidth / 2;
  const startPoint = `${linecap === "round" ? YCoordOfStartPoint : 0}, ${YCoordOfStartPoint}`;
  const endPoint = `${linecap === "round" ? YCoordOfEndPoint : 100}, ${YCoordOfStartPoint}`;
  const pathString = `M ${startPoint} L ${endPoint}`;
  const viewBoxString = `0 0 100 ${lineWidth}`;

  const pathStyle = {
    strokeDasharray: `${percent}px, 100px`,
    strokeDashoffset: "0px",
    transition: "stroke-dashoffset 0.3s ease 0s, stroke-dasharray 0.3s ease 0s, stroke 0.3s linear",
  };

  const wrapClassNames = classNames(
    `${prefixCls}-line-wrapper`,
    `${prefixCls}-line-${status}`,
    className,
  );

  return (
    <LineWrapper
      lineWidth={lineWidth}
      prefixCls={prefixCls}
      className={wrapClassNames}
      style={wrapStyle}
    >
      <svg
        className={`${prefixCls}-line`}
        viewBox={viewBoxString}
        preserveAspectRatio="none"
        style={style}
      >
        <path
          className={`${prefixCls}-line-trail`}
          d={pathString}
          strokeLinecap={linecap}
          stroke={trailColor}
          strokeWidth={lineWidth}
          fillOpacity="0"
        />
        <path
          className={`${prefixCls}-line-path`}
          d={pathString}
          strokeLinecap={linecap}
          stroke={adoptStrokeColor(percent, status)}
          strokeWidth={lineWidth}
          fillOpacity="0"
          style={pathStyle}
        />
      </svg>
      {showInfo && (
        <span className={`${prefixCls}-info`}>{adoptInfoContent(percent, status, format)}</span>
      )}
    </LineWrapper>
  );
};

export default Line;
