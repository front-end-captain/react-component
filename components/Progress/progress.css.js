import styled from "styled-components";

export const LineWrapper = styled.div`
  display: flex;
  width: 100%;
  align-items: center;

  .${(props) => props.prefixCls}-line {
    flex: 9 1;
  }

  .${(props) => props.prefixCls}-info {
    display: flex;
    flex: 1 1;
    word-break: normal;
    white-space: nowrap;
    font-size: 14px;
    line-height: 1;
    padding-left: 14px;
    color: rgba(0, 0, 0, 0.45);
  }
`;

export const CircleWrapper = styled.div`
  display: inline-block;
  position: relative;
  width: 120px;
  height: 120px;

  .${(props) => props.prefixCls}-info {
    position: absolute;
    width: 100%;
    text-align: center;
    line-height: 1;
    font-size: 24px;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    margin: 0;
    padding: 0;
    color: rgba(0, 0, 0, 0.65);
    white-space: normal;
  }
`;
