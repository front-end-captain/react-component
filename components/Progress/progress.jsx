import React from "react";
import PropTypes from "prop-types";

import Line from "./line.jsx";
import Circle from "./circle.jsx";
import { ErrorIcon, SuccessIcon } from "./../Icon/index.jsx";

import { progressTrailColor, successColor, errorColor, infoColor } from "./../_style/vars.js";

const Progress = (props) => {
  const { type, lineColor } = props;

  const adoptStrokeColor = (percent, status) => {
    if (percent === 0) {
      return progressTrailColor;
    }

    if (status === "exception" && percent !== 0) {
      return errorColor;
    }
    if (status === "success" || percent === 100) {
      return successColor;
    }

    return lineColor;
  };

  const adoptInfoContent = (percent, status, formatMethod) => {
    if (formatMethod !== undefined) {
      return formatMethod(percent);
    }
    if (status === "exception") {
      return <ErrorIcon style={{ color: errorColor }} />;
    }
    if (status === "success" || percent === 100) {
      return <SuccessIcon style={{ color: successColor }} />;
    }

    return `${percent}%`;
  };

  const adoptMethod = { adoptInfoContent, adoptStrokeColor };

  const mixProps = { ...props, ...adoptMethod, trailColor: progressTrailColor };

  if (type === "circle" || type === "dashboard") {
    return <Circle {...mixProps} />;
  }

  return <Line {...mixProps} />;
};

Progress.defaultProps = {
  prefixCls: "sky-progress",
  type: "line",
  percent: 0,
  showInfo: true,
  lineColor: infoColor,
  linecap: "round",
  lineWidth: 1.5,
  status: "normal",
};

Progress.propTypes = {
  prefixCls: PropTypes.string,
  className: PropTypes.string,
  type: PropTypes.oneOf(["line", "circle", "dashboard"]),
  percent: PropTypes.number,
  showInfo: PropTypes.bool,
  status: PropTypes.oneOf(["normal", "exception", "success"]),
  linecap: PropTypes.oneOf(["round", "square"]),
  lineColor: PropTypes.string,
  lineWidth: PropTypes.number,
  style: PropTypes.objectOf(PropTypes.string),
  wrapStyle: PropTypes.objectOf(PropTypes.string),
  format: PropTypes.func,
};

export default Progress;
