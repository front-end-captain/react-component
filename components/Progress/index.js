import Progress from "./progress.jsx";

export default Progress;
