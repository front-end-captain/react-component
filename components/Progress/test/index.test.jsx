import React from "react";
import assert from "power-assert";
import { render } from "enzyme";
import toJson from "enzyme-to-json";

import Progress from "./../index.js";

describe("<Progress />", () => {
  it("should be render some <Progress /> components", () => {
    const wrapper = render(
      <div>
        <Progress />
        <Progress type="dashboard" />
        <Progress type="circle" />
        <Progress status="success" percent={100} />
      </div>,
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should find some class names", () => {
    const wrapper = render(
      <div>
        <Progress />
        <Progress type="dashboard" />
        <Progress type="circle" />
        <Progress status="success" percent={100} />
        <Progress status="exception" percent={45} />
      </div>,
    );

    assert(wrapper.find(".sky-progress-line-wrapper").length === 3);
    assert(wrapper.find(".sky-progress-circle-wrapper").length === 2);
    assert(wrapper.find(".sky-progress-line-success").length === 1);
    assert(wrapper.find(".sky-progress-line-normal").length === 1);
    assert(wrapper.find(".sky-progress-circle-normal").length === 2);
    assert(wrapper.find(".sky-progress-line-exception").length === 1);
  });
});
