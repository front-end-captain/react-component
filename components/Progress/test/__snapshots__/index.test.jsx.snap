// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Progress /> should be render some <Progress /> components 1`] = `
<div>
  <div
    class="sky-progress-line-wrapper sky-progress-line-normal progresscss__LineWrapper-qlii17-0 dJdwwr"
  >
    <svg
      class="sky-progress-line"
      preserveAspectRatio="none"
      viewBox="0 0 100 1.5"
    >
      <path
        class="sky-progress-line-trail"
        d="M 0.75, 0.75 L 99.25, 0.75"
        fill-opacity="0"
        stroke="#f5f5f5"
        stroke-linecap="round"
        stroke-width="1.5"
      />
      <path
        class="sky-progress-line-path"
        d="M 0.75, 0.75 L 99.25, 0.75"
        fill-opacity="0"
        stroke="#f5f5f5"
        stroke-linecap="round"
        stroke-width="1.5"
        style="stroke-dasharray:0px, 100px;stroke-dashoffset:0px;transition:stroke-dashoffset 0.3s ease 0s, stroke-dasharray 0.3s ease 0s, stroke 0.3s linear"
      />
    </svg>
    <span
      class="sky-progress-info"
    >
      0%
    </span>
  </div>
  <div
    class="sky-progress-circle-wrapper sky-progress-circle-normal progresscss__CircleWrapper-qlii17-1 fUyPWZ"
  >
    <svg
      class="sky-progress-circle"
      viewBox="0 0 100 100"
    >
      <path
        class="sky-progress-circle-trail"
        d="M 50,50 m 0,47.75
     a 47.75,47.75 0 1 1 0,-95.5
     a 47.75,47.75 0 1 1 0,95.5"
        fill-opacity="0"
        stroke="#f5f5f5"
        stroke-linecap="round"
        stroke-width="4.5"
        style="stroke-dasharray:230.02209841782525px 300.02209841782525px;stroke-dashoffset:-35px;transition:stroke-dashoffset 0.3s ease 0s, stroke-dasharray 0.3s ease 0s, stroke 0.3s, stroke-width 0.06s ease 0.3s"
      />
      <path
        class="sky-progress-circle-path"
        d="M 50,50 m 0,47.75
     a 47.75,47.75 0 1 1 0,-95.5
     a 47.75,47.75 0 1 1 0,95.5"
        fill-opacity="0"
        stroke="#f5f5f5"
        stroke-linecap="round"
        stroke-width="0"
        style="stroke-dasharray:0px 300.02209841782525px;stroke-dashoffset:-35px;transition:stroke-dashoffset 0.3s ease 0s, stroke-dasharray 0.3s ease 0s, stroke 0.3s, stroke-width 0.06s ease 0.3s"
      />
    </svg>
    <span
      class="sky-progress-info"
    >
      0%
    </span>
  </div>
  <div
    class="sky-progress-circle-wrapper sky-progress-circle-normal progresscss__CircleWrapper-qlii17-1 fUyPWZ"
  >
    <svg
      class="sky-progress-circle"
      viewBox="0 0 100 100"
    >
      <path
        class="sky-progress-circle-trail"
        d="M 50,50 m 0,-47.75
     a 47.75,47.75 0 1 1 0,95.5
     a 47.75,47.75 0 1 1 0,-95.5"
        fill-opacity="0"
        stroke="#f5f5f5"
        stroke-linecap="round"
        stroke-width="4.5"
        style="stroke-dasharray:300.02209841782525px 300.02209841782525px;stroke-dashoffset:-0px;transition:stroke-dashoffset 0.3s ease 0s, stroke-dasharray 0.3s ease 0s, stroke 0.3s, stroke-width 0.06s ease 0.3s"
      />
      <path
        class="sky-progress-circle-path"
        d="M 50,50 m 0,-47.75
     a 47.75,47.75 0 1 1 0,95.5
     a 47.75,47.75 0 1 1 0,-95.5"
        fill-opacity="0"
        stroke="#f5f5f5"
        stroke-linecap="round"
        stroke-width="0"
        style="stroke-dasharray:0px 300.02209841782525px;stroke-dashoffset:-0px;transition:stroke-dashoffset 0.3s ease 0s, stroke-dasharray 0.3s ease 0s, stroke 0.3s, stroke-width 0.06s ease 0.3s"
      />
    </svg>
    <span
      class="sky-progress-info"
    >
      0%
    </span>
  </div>
  <div
    class="sky-progress-line-wrapper sky-progress-line-success progresscss__LineWrapper-qlii17-0 dJdwwr"
  >
    <svg
      class="sky-progress-line"
      preserveAspectRatio="none"
      viewBox="0 0 100 1.5"
    >
      <path
        class="sky-progress-line-trail"
        d="M 0.75, 0.75 L 99.25, 0.75"
        fill-opacity="0"
        stroke="#f5f5f5"
        stroke-linecap="round"
        stroke-width="1.5"
      />
      <path
        class="sky-progress-line-path"
        d="M 0.75, 0.75 L 99.25, 0.75"
        fill-opacity="0"
        stroke="#52c41a"
        stroke-linecap="round"
        stroke-width="1.5"
        style="stroke-dasharray:100px, 100px;stroke-dashoffset:0px;transition:stroke-dashoffset 0.3s ease 0s, stroke-dasharray 0.3s ease 0s, stroke 0.3s linear"
      />
    </svg>
    <span
      class="sky-progress-info"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        style="color:#52c41a"
        viewBox="0 0 512 512"
        width="1em"
      >
        <path
          d="M256 8C119.033 8 8 119.033 8 256s111.033 248 248 248 248-111.033 248-248S392.967 8 256 8zm0 48c110.532 0 200 89.451 200 200 0 110.532-89.451 200-200 200-110.532 0-200-89.451-200-200 0-110.532 89.451-200 200-200m140.204 130.267l-22.536-22.718c-4.667-4.705-12.265-4.736-16.97-.068L215.346 303.697l-59.792-60.277c-4.667-4.705-12.265-4.736-16.97-.069l-22.719 22.536c-4.705 4.667-4.736 12.265-.068 16.971l90.781 91.516c4.667 4.705 12.265 4.736 16.97.068l172.589-171.204c4.704-4.668 4.734-12.266.067-16.971z"
        />
      </svg>
    </span>
  </div>
</div>
`;
